text_generation/causal_lm/cpp/continuous_batching/library/src/scheduler.hpp

```diff
@@ -89,54 +90,84 @@ class Scheduler {
         return num_running;
     }
 
-    void _preempt_by_recompute(SequenceGroup::Ptr sequence_group) {
+
+    bool _preempt_by_recompute(SequenceGroup::Ptr sequence_group, size_t blocks_needed) {
+        size_t total_num_released_blocks = 0;
+        size_t processed_tokens = sequence_group->get_num_processed_tokens();
+        size_t block_size = m_config.block_size;
+        size_t prev_blocks_count = m_block_manager.num_free_blocks();
+        size_t preempted_tokens = 0;
+
+
         // currently, we support only preemption by (TODO: implement "partial") recompute
         for (size_t s = 0; s < sequence_group->num_running_seqs(); ++s) {
-            // so, let's fully drop a sequence(s) from block_manager
-            m_block_manager.free_sequence((*sequence_group)[s]->get_id());
+            auto seq_id = (*sequence_group)[s]->get_id();
+            if (!m_block_manager.has_block_table(seq_id)) {
+                // no blocks are allocated for this sequence, so it can't be preempted
+                return false;
+            }
+            auto block_table = m_block_manager.get_block_table(seq_id);
+            size_t required_blocks = blocks_needed - total_num_released_blocks;
+            if (required_blocks >= block_table.size()) {
+                // fully drop a sequence(s) from block_manager
+                m_block_manager.free_sequence(seq_id);
+            }
+            else {
+                m_block_manager.free_sequence_partially(seq_id, required_blocks);
+            }
+
+            // calculate the number of released blocks
+            auto released_blocks = m_block_manager.num_free_blocks() - prev_blocks_count;
+            total_num_released_blocks += released_blocks;
+            prev_blocks_count = m_block_manager.num_free_blocks();
+
+
+            // calculate the number of preempted tokens
+            auto tokens_in_last_block = processed_tokens % block_size;
+            if (tokens_in_last_block == 0) {    
+                tokens_in_last_block = block_size;
+            }
+
+            preempted_tokens += tokens_in_last_block + std::max<size_t>((int)released_blocks - 1, 0) * block_size;
+            if (m_block_manager.num_free_blocks() >= blocks_needed) {
+                break;
```

Comment:
looks like we can have cases when sequences within sequence group have different context_lens / state, right?
It will be invalid case, because they should be aligned to be properly scheduled on future iterations.

I suppose we can keep it as future improvement.

Reply:
Yes, this logic works only for single sequence case. Added logic to run full sequence preemption for non-gready sampling and TODO for implementing multi-sequence case.